Toshi/Source/Toshi/Core/TClass.h

```diff
@@ -3,32 +3,59 @@
 
 namespace Toshi
 {
+	// TClassProps definitions
 	typedef void* (*t_CreateTObject)();
-	typedef void* (*t_CreateTObjectAtPlace)(void*);
-	typedef void  (*t_RegisterScriptingAPI)();
+	typedef void* (*t_CreateTObjectInPlace)(void*);
+	typedef void  (*t_InitializeStatic)();
+	typedef void  (*t_UninitializeStatic)();
 
-	class TClass
+	// RecurseTree definitions
+	typedef void  (*t_ResurceTreeCheck)(class TClass*);
+	typedef void  (*t_ResurceTreeAction)(class TClass*);
+
+	// This is a separate struct because we want to do compile-time initialization
+	struct TClassProps
+	{
+		const char* m_Name = 0;									// 0x00
+		t_CreateTObject m_Create = 0;							// 0x04
+		t_CreateTObjectInPlace m_CreateInPlace = 0;				// 0x08
+		t_InitializeStatic m_Initialize = 0;					// 0x0C
+		t_UninitializeStatic m_Uninitialize = 0;				// 0x10
+		TClassProps* m_Parent = 0;								// 0x14
+		TClassProps* m_Previous = 0;							// 0x18
+		TClassProps* m_LastAttached = 0;						// 0x1C
+		uint32_t m_Version = 0;									// 0x20
+		size_t m_Size = 0;										// 0x24
+		uint32_t m_Unk = 0;										// 0x28
+		bool m_Initialized = 0;									// 0x2C
+	};
+
+	class TClass : public TClassProps
 	{
 	public:
-		const char* m_pcClassName; //0
-		t_CreateTObject m_pCreateTObject; //0x4
-		t_CreateTObjectAtPlace m_pCreateTObjectAtPlace; //0x8
-		t_RegisterScriptingAPI m_pRegisterScriptingAPI; //0xC
-		void* m_pFunc; //0x10
-
-		TClass* m_parent; //0x14
-		TClass* m_attached; //0x18
-		TClass* m_tclass; //0x1C
-		size_t m_size; //0x20
-		uint32_t m_version; // 0x24
-		void* m_unk4; // 0x28
-		bool m_bInitialised; // 0x2C
-
-		TClass(const char* name, TClass* parent, t_CreateTObject create, t_CreateTObjectAtPlace createAtPlace, t_RegisterScriptingAPI scripting, uint16_t verMajor, uint16_t verMinor, size_t size);
-
-		inline const char* GetName() const { return m_pcClassName; }
-		bool AttachClassToParent();
-		bool IsAttached() const;
+		TClass(const char* name, TClassProps* parent, uint32_t version, size_t size, t_CreateTObject fCreate, t_CreateTObjectInPlace m_CreateInPlace, t_InitializeStatic fInit, t_UninitializeStatic fUninit);
+		~TClass();
+
+		void Initialize();
+		void RecurseTree();
+		void RecurseTree2();
+
+		inline bool IsInitialized() { return m_Initialized; }
+		inline const char* GetName() const { return m_Name; }
+		inline TClassProps& GetPropsRef() { return *this; }
+		inline TClassProps* GetProps() { return this; }
+
+		// todo: move this function away from this class
+		static bool TryInitialize(TClass* tClass);
 	};
+
+	// TClass should be equal to TClassProps
+	static_assert(sizeof(TClass) == sizeof(TClassProps));
```

Comment:
TAssert?

Reply:
static_assert works at the moment of compilation, TASSERT works at runtime